src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.Helpers.cs

```diff
@@ -23,7 +25,21 @@ private static void ValidateInputOutputSpanNonOverlapping<T>(ReadOnlySpan<T> inp
         }
 
         /// <summary>Throws an <see cref="OverflowException"/> for trying to negate the minimum value of a two-complement value.</summary>
-        internal static void ThrowNegateTwosCompOverflow() => throw new OverflowException(SR.Overflow_NegateTwosCompNum);
+        private static void ThrowNegateTwosCompOverflow() => throw new OverflowException(SR.Overflow_NegateTwosCompNum);
+
+        /// <summary>Creates a span of <typeparamref name="TTo"/> from a <typeparamref name="TFrom"/> when they're the same type.</summary>
+        private static unsafe Span<TTo> Rename<TFrom, TTo>(Span<TFrom> span) // MemoryMarshal.Cast, but without constraints or validation
+        {
+            Debug.Assert(sizeof(TFrom) == sizeof(TTo));
+            return MemoryMarshal.CreateSpan(ref Unsafe.As<TFrom, TTo>(ref MemoryMarshal.GetReference(span)), span.Length);
```

Comment:
```suggestion
            Debug.Assert(typeof(TFrom) == typeof(TTo));
            return *(Span<TTo>*)&span;
```

Reply:
also, this is always inlined on .NET Framework

Reply:
> `return *(Span<TTo>*)&span;`

This is actually [a GC hole on x64](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsACgylWoxbtOASRYiIfcVCkz5y4gCYqy8wGY1Scs1zYAZjEoXyAFVkZlAN7K5CGU9tSOonzYTAA8roIQ2BgAfOQAggAUUTGxrCxpuACUwaFBSqGVlADs5ABU2dFxCUmpdUUAZGYVoQC+pSHE4VSRTfGJyWkAQo25+ankxQPk5VWhxLUAsjD40ACem9hQuAAW2ILcsMkwOUyZsK7kAKpMLu406bh5LGjkLZP3GCPba7KAHI6nc40ADiMAwdCBIhgTBkmWKRV+uBoABlkQBzDAnIqKHohfpKXpAA===);

Reply:
@MichalPetryka Is this a GC hole when `typeof(TFrom) == typeof(TTo)`?

Reply:
> different layouts for different spans

These aren't different spans. They're the exact same type.

Reply:
> > different layouts for different spans
> 
> These aren't different spans. They're the exact same type.

Ah right, I've read that as `sizeof`, not `typeof`. Unity IL2CPP actually has funny cases where one of the Spans could be a fast span and the other a portable one - if you think that's a bug, they closed my issue report as by design.

Reply:
Presumably the proposed patch will require extra +2 lines to suppress a warning so IMO the current code is fine.

Reply:
> Ah right, I've read that as sizeof, not typeof

It should actually be typeof. I'll change that.

Reply:
> Presumably the proposed patch will require extra +2 lines to suppress a warning so IMO the current code is fine.

The current code fails to compile on `netstandard2.0` due to lack of `MemoryMarshal.CreateSpan`. This suggestion would resolve this.

Reply:
> > Ah right, I've read that as sizeof, not typeof
> 
> It should actually be typeof. I'll change that.

@stephentoub If you don't go with this suggestion then the `pragma warning disable CS8500` can be removed.